lib/webdriveragent.js

```diff
@@ -287,6 +289,70 @@ class WebDriverAgent {
     }
   }
 
+
+  /**
+   * @typedef {Object} launchWdaViaDeviceCtlOptions
+   * @property {Record<string, string|number>} env environment variables for the launching WDA process
+   */
+
+  /**
+   * Launch WDA with preinstalled package with 'xcrun devicectl device process launch'.
+   * The WDA package must be prepared properly like published via
+   * https://github.com/appium/WebDriverAgent/releases
+   * with proper sign for this case.
+   *
+   * When we implement launching XCTest service via appium-ios-device,
+   * this implementation can be replaced with it.
+   *
+   * @param {?launchWdaViaDeviceCtlOptions} opts launching WDA with devicectl command options.
+   * @return {Promise<void>}
+   */
+
+  async _launchWdaViaDeviceCtl(opts) {
+    // FIXME: use appium-xcuitest-driver's Devicectl. Maybe it needs to be included in the `this.device`?
+    //
+
+    const {env} = opts || {};
+
+    let xcrunBinnaryPath;
+    try {
+      xcrunBinnaryPath = await fs.which('xcrun');
+    } catch (e) {
+      throw new Error(
+        `xcrun has not been found in PATH. ` +
+          `Please make sure XCode development tools are installed`,
+      );
+    }
+
+    const {stdout} = await exec(xcrunBinnaryPath, [
+      'devicectl',
+      'device',
+      'process',
+      'launch',
+      `--device`, this.device.udid,
+      '--environment-variables', JSON.stringify(_.mapValues(env, (v) => _.toString(v))),
+      '--terminate-existing',
+      this.bundleIdForXctest
+    ]);
+    this.log.debug(`The output of devicectl command: ${stdout}`);
+
+    // Launching app via decictl does not wait for the app start.
+    // We should wait for the app start by ourselves.
+    try {
+      await waitForCondition(async () => !_.isNull(await this.getStatus()), {
```

Comment:
we should wrap getStatus into try/catch if it is supposed to throw an exception. The waitForCondition helper passes all exceptions through

Reply:
could below method raise an error...? I thought `return await noSessionProxy.command('/status', 'GET');` actually could be, but it had try/catch and return null then.
```
  async getStatus () {
    const noSessionProxy = new NoSessionProxy({
      server: this.url.hostname,
      port: this.url.port,
      base: this.basePath,
      timeout: 3000,
    });
    try {
      return await noSessionProxy.command('/status', 'GET');
    } catch (err) {
      this.log.debug(`WDA is not listening at '${this.url.href}'`);
      return null;
    }
  }
```